monkey/monkey_island/cc/version.py

```diff
@@ -55,29 +53,26 @@ def download_url(self):
         return self._download_url
 
     def _set_version_metadata(self):
-        self._latest_version = self._get_latest_version()
-        self._download_url = self._get_download_link()
+        self._latest_version, self._download_url = self._get_version_info()
         self._initialization_complete.set()
 
-    def _get_latest_version(self) -> str:
-        url = VERSION_SERVER_CHECK_NEW_URL % (self._deployment.value, self._version_number)
+    def _get_version_info(self) -> Tuple[Optional[str], Optional[str]]:
+        url = LATEST_VERSION_URL.format(self._deployment.value)
 
         try:
-            reply = requests.get(url, timeout=LATEST_VERSION_TIMEOUT)
+            response = requests.get(url, timeout=LATEST_VERSION_TIMEOUT).json()
         except requests.exceptions.RequestException as err:
-            logger.warning(f"Failed to connect to {VERSION_SERVER_URL_PREF}: {err}")
-            return self._version_number
-
-        res = reply.json().get("newer_version", None)
-
-        if res is False:
-            return self._version_number
+            logger.warning(f"Failed to connect to {url}, Error: {err}")
+            return self._version_number, None
```

Comment:
This is a defect. The responsibility of `self.latest_version` is to represent the latest version. In this case the latest version is unknown, not the current version. The "latest version some component could fetch" is not a thing that exists in the domain, so it doesn't need to be represented (it's also an oxymoron because it couldn't fetch anything). Since the responsibility of this component is to gather the data, not consume it, making decisions like this here is wrong.

Reply:
> The responsibility of self.latest_version is to represent the latest version.

Slight tweak: latest _known_ version

> In this case the latest version is unknown

No, the latest version that the Island knows about is not unknown.

> The "latest version some component could fetch" is not a thing that exists in the domain

The component is not defined as "the latest version I could fetch". It's "the latest version I know about". Using `None` to mean "failed to contact server" is not only opaque and unhelpful to the user, but it leaks implementation details through the interface.

> so it doesn't need to be represented (it's also an oxymoron because it couldn't fetch anything). 

Whether or not you were able to successfully contact the server, the latest known version can't be `None`. That's not a thing that exists in the domain (unless the software doesn't exist). It's an oxymoron because you're saying there is no latest version, but of course there is. To the best of the Island's knowledge, the latest version is the one you're using.

> Since the responsibility of this component is to gather the data

The responsibility of the component is not to gather data. It's responsibility is to represent the latest known version. How it comes about that decision is an implementation detail, and shouldn't be leaked to the consumers.